### PR TITLE
Symmetry analysis

### PR DESCRIPTION
**Pull request by imagoulas**
## Description
Added a symmetry analysis function. For a given quantum state, it computes the expectation values and variances of the particle number N, projection of total spin on the z axis Sz, and total spin squared S^2 operators. It also computes the weight contribution to the wavefunction of the various N, Sz, and spatial symmetry sectors.

A test of this functionality will be included after the implementation of the approximate FEB/QEB circuits, which break N and Sz symmetries.

In an initial commit, I accidentally included modifications to the code pertaining to the energy shift of Moller-Plesset denominators in Jacobi iterations. These additions have been removed.

## User Notes
- [x ] Features added
- [ ] Changes to compilation (if any)

## Checklist
- [ ] Added/updated tests of new features
- [ ] Removed comments in input files
- [x ] Documented source code
- [ ] Checked for redundant headers/imports
- [ ] Checked for consistency in the formatting of the output file
- [x ] Ready to go!

### CODE REVIEW

File: src/qforte/utils/symmetry_analysis.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        List that holds the irreps of the group in the Cotton ordering.
+
+    orb_irreps_to_int: list of int
+        Integer n indexes the irrep of spatial orbital n.
+    """
+
+    # Generate the quantum state defined by the Uprep, tops, tamps, compact, qubit and trotter_number
+    qc = qf.Computer(n_qubits)
+    qc.apply_circuit(Uprep)
+    if compact:
```
Comment (JonathonMisiewicz) [inline src/qforte/utils/symmetry_analysis.py:49]: I _really_ don't like that this code needs to know the details of how to construct the computer. Can we have the code take in the Computer as an argument, instead?

Comment (imagoulas) [reply src/qforte/utils/symmetry_analysis.py:49]: I do agree with you that the code looks ugly. If we pass the Computer as an argument, whenever we want to use this function we will need to do the following:
```
U = self.ansatz_circuit()
comp = qf.Computer(self._nqb)
comp.apply_circuit(self._Uprep)
comp.apply_circuit(U)
```
Indeed, much more concise. Should I proceed with the simplifications?

Comment (JonathonMisiewicz) [reply src/qforte/utils/symmetry_analysis.py:49]: You can make it even shorter.
```
comp = qf.Computer(self._nqb)
comp.apply_circuit(self.build_Uvqc())
```

Yes, simplify ahead.

File: src/qforte/utils/symmetry_analysis.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    print('\n\n*******************')
+    print('*Symmetry Analysis*')
+    print('*******************')
+
+    print('\nParticle number symmetry analysis')
+    print('=================================')
+
+    for i in range(n_qubits + 1):
+        print('N =', str(i).rjust(2, ' '), 'weight:', f'{weight_particle_number[i]:.10f}')
```
Comment (JonathonMisiewicz) [inline src/qforte/utils/symmetry_analysis.py:109]: What are these "weights", and why do we print them on a per-qubit basis?

Comment (imagoulas) [reply src/qforte/utils/symmetry_analysis.py:109]: The weights quantify the contribution of each symmetry sector to the final wavefunction. For example, let us assume that the wavefunction of a 6-electron (N = 6) system looks like this:

|Ψ> = c_1 |Φ_1> + c_2 |Φ_2> + c_3 |Φ_3> + c_4 |Φ_4> + c_5 |Φ_5> + c_6 |Φ_6> + d_1 |D_1> + d_2 |D_2>,

where  the last two determinants are contaminants containing N = 5 electrons. We thus have the following weights:

weight(N = 5) =  |d_1|^2 + |d_2|^2
weight(N = 6) =  |c_1|^2 + |c_2|^2 + |c_3|^2 + |c_4|^2 + |c_5|^2 + |c_6|^2

Normalization requires:

weight(N = 5) + weight(N = 6) = 1

They are not printed per qubit, but per total particle number, N in [0, n_qubits].

The code works in a similar manner for the Sz and spatial symmetries. Since we are using a basis of Slater determinants, it is not straightforward to do a similar analysis for S^2.

File: src/qforte/utils/symmetry_analysis.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        weight = qc.get_coeff_vec()[det]
+        weight = weight * np.conjugate(weight)
+        particle_number = len(occ_alpha + occ_beta)
+        Sz_value = (len(occ_alpha) - len(occ_beta)) * 0.5
+        irrep = qf.sq_op_find_symmetry(orb_irreps_to_int, occ_alpha + occ_beta, [])
+        weight_particle_number[particle_number] += weight
+        weight_Sz[Sz_values_dictionary[Sz_value]] += weight
+        weight_irreps[irrep] += weight
+    for i in range(n_qubits + 1):
+        weight_particle_number[i] = np.real(weight_particle_number[i])
```
Comment (JonathonMisiewicz) [inline src/qforte/utils/symmetry_analysis.py:96]: l. 88 should guarantee these are real, so why is this necessary? At _most_, just call `np.real` on l. 88.

File: src/qforte/utils/symmetry_analysis.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    S_sqrd_sqrd.add_op(S_sqrd)
+    S_sqrd_sqrd.operator_product(S_sqrd, True, True)
+    S_sqrd_exp_val = qc.direct_op_exp_val(S_sqrd)
+    S_sqrd_var = qc.direct_op_exp_val(S_sqrd_sqrd) - S_sqrd_exp_val**2
+    if np.imag(S_sqrd_exp_val) !=0 or np.imag(S_sqrd_var) !=0:
+        raise ValueError('The expectation value and variance of S^2 should be real!')
+    S_sqrd_exp_val = np.real(S_sqrd_exp_val)
+    S_sqrd_var = np.real(S_sqrd_var)
+
+    # Compute the contributions of various symmetry spaces to the wavefunction
```
Comment (JonathonMisiewicz) [inline src/qforte/utils/symmetry_analysis.py:70]:
```suggestion
    # Compute the weight of various symmetry subspaces in the wavefunction
```

File: src/qforte/utils/symmetry_analysis.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    weight_particle_number = [0] * (n_qubits + 1)
```
Comment (JonathonMisiewicz) [inline src/qforte/utils/symmetry_analysis.py:71]: I'd rename this variable to `weight_per_particle_number`, and so forth.

**Review by JonathonMisiewicz — APPROVED** [review]

**Comment by JonathonMisiewicz** [comment]
Approval here means, "fine to merge after addressing comments."

File: src/qforte/utils/symmetry_analysis.py (hunk 3 above)
Comment (imagoulas) [reply src/qforte/utils/symmetry_analysis.py:96]: This is done purely for printing purposes: I want to avoid printing 0.0000000000j all the time. I will follow your suggestion and call `np.real` on line 88.
